### PR TITLE
Add type defs for styled-components v3.x.x

### PR DESCRIPTION
**Pull request by nvie**
This is literally a copy of the v2.x.x files.

### CODE REVIEW

File: definitions/npm/styled-components_v3.x.x/flow_v0.53.x-/styled-components_v3.x.x.js [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+>($npm$styledComponents$ReactComponentFunctionalUndefinedDefaultProps<InputProps>) => $npm$styledComponents$ReactComponentFunctional<OutputProps, { theme: $npm$styledComponents$Theme }>
+
+type $npm$styledComponents$WithTheme =
+  & $npm$styledComponents$WithThemeReactComponentClass
+  & $npm$styledComponents$WithThemeReactComponentClassUndefinedDefaultProps
+  & $npm$styledComponents$WithThemeReactComponentFunctional
+  & $npm$styledComponents$WithThemeReactComponentFunctionalUndefinedDefaultProps
+
+// ---- MISC ----
+type $npm$styledComponents$Theme = {[key: string]: mixed};
```
Comment (k15a) [inline definitions/npm/styled-components_v3.x.x/flow_v0.53.x-/styled-components_v3.x.x.js]: Can we change this to `Object`? It's not working probably when you use a nested theme.

**Comment by dashed** [comment]
@nvie any progress on this?

**Comment by nvie** [comment]
Not sure what the expected process is, so no. All I did here was copy over the libdef files from v2 to v3 since most of the API surface has not been changed, practically. I'm not involved in the styled-components project, so I don't have a good idea what has changed from v2 -> v3. If I run the libdefs tests locally, I get a ton of daunting errors which I don't understand because I'm unfamiliar with the internals, even for the styled-component_v2.

For the projects I'm using styled-components for, I've simply copied the v2 libdefs and not hit an issue with them.

**Comment by AndrewSouthpaw** [comment]
Hm. There's only four errors, would someone be willing to look into fixing them? I'd much rather not push in a new libdef that's already failing.

**Comment by dashed** [comment]
@nvie are you still looking to resolve these errors? Do you not mind if I try to tackle them?

**Comment by nvie** [comment]
> Do you not mind if I try to tackle them?

I don't mind. In fact, the opposite — I'd appreciate it very much 😄

**Comment by zalishchuk** [comment]
What's the problem right now, why this can not be fixed for a very long time?

**Comment by saadq** [comment]
If no one else is currently working on this, I'll look into fixing this tomorrow.

**Comment by dashed** [comment]
@saadq please do 👍 I wasn't able to find time to investigate this. 😞

**Comment by saadq** [comment]
Hey I just took a look at this. It seems like the errors only occur on versions 56 and below. @GAntoine and @AndrewSouthpaw what are your thoughts on pushing a library definition that is only `0.57.x-`? 

If that's okay, is there a proper way for me to fix this PR? I'm pretty noob with Git, so I was just thinking about opening a new Pull Request but not sure if there's a better way.

**Comment by dashed** [comment]
@nvie @saadq I've somewhat looked into this and opened a PR at https://github.com/flowtype/flow-typed/pull/2046

**Comment by GAntoine** [comment]
@nvie With #2046 merged in, would you be able to rebase this on the latest master?

**Comment by nvie** [comment]
Perfect! Thanks @dashed + @GAntoine!

**Pull request closed without merging**